app/server/Main.hs

```diff
@@ -1,6 +1,81 @@
+{-# LANGUAGE QuasiQuotes #-}
+
+{-
+   This module parses command line arguments and starts the wallet.
+
+   This should be the only module that has a notion of command line arguments.
+   Although, for development purposes, we may extend the docopt specification
+   here and call `getArgs` where it is needed.
+-}
 module Main where
 
+import Control.Monad
+    ( when )
+import Data.Maybe
+    ( fromMaybe )
 import Prelude
+import System.Console.Docopt
+    ( Docopt
+    , docopt
+    , exitWithUsage
+    , getArgOrExitWith
+    , isPresent
+    , longOption
+    , parseArgsOrExit
+    )
+import System.Environment
+    ( getArgs )
+import Text.Read
+    ( readMaybe )
+
+
+-- | Command-Line Interface specification. See http://docopt.org/
+cli :: Docopt
+cli = [docopt|
+cardano-wallet-server
+
+Start the cardano wallet along with its API and underlying node.
+
+Usage:
+  cardano-wallet-server [options]
+  cardano-wallet-server --help
+
+Options:
+  --wallet-server-port <PORT>  port used for serving the wallet API [default: 8090]
+  --node-port <PORT>           port used for node-wallet communication [default: 8080]
+  --network <NETWORK>          mainnet or testnet [default: mainnet]
+|]
+
+
+data Network = Mainnet | Testnet
+    deriving (Show, Enum)
 
 main :: IO ()
-main = return ()
+main = do
+    args <- parseArgsOrExit cli =<< getArgs
+    when (args `isPresent` (longOption "help")) $ exitWithUsage cli
+
+    let get option = args `getArgOrExit` option
+
+    network <- readNetwork =<< get (longOption "network")
+    walletPort <- readInt <$> get (longOption "wallet-server-port")
+    nodePort <- readInt <$> get (longOption "node-port")
+
+    putStrLn $
+        "TODO: start wallet on port " ++ (show walletPort) ++
+        ",\n      connecting to " ++ (show network) ++
+        " node on port " ++ (show nodePort)
+  where
+    readInt :: String -> Int
+    readInt s = fromMaybe (error $ "not an integer: " ++ show s) (readMaybe s)
+
+    getArgOrExit = getArgOrExitWith cli
+
+    readNetwork :: String -> IO Network
```

Comment:
This function conflates two responsibilities: decoding a value, and reporting an error to the user. I'd expect such function to actually be pure and return a `Maybe Network` or `Either Err|Text Network`.

What about defining a helper function as:

```hs
    getArg
        :: Arguments
        -> Option
        -> (String -> Either String a)
        -> IO a
    getArg args opt decode = do
        str <- getArgOrExitWith cli args opt
        case decode str of
            Right a -> return a
            Left err -> do
                putStrLn $ "Invalid '" <> show opt <> "' (" <> err <> ")"
                exitWithUsage cli
```

and have decoders as `readXXX :: String -> Either String XXX` ? This way, we separate running the parsers from the parsers themselves.